modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotCheckDistributedProcess.java

```diff
@@ -0,0 +1,549 @@
+
+    /** */
+    private final IgniteLogger log;
+
+    /** */
+    private final GridKernalContext kctx;
+
+    /** Snapshot check requests per snapshot on every node. */
+    final Map<String, SnapshotCheckProcessRequest> requests = new ConcurrentHashMap<>();
+
+    /** Cluster-wide operation futures per snapshot called from current node. */
+    private final Map<UUID, GridFutureAdapter<SnapshotPartitionsVerifyTaskResult>> clusterOpFuts = new ConcurrentHashMap<>();
+
+    /** Check metas first phase subprocess. */
+    private final DistributedProcess<SnapshotCheckProcessRequest, ArrayList<SnapshotMetadata>> phase1CheckMetas;
+
+    /** Partition hashes second phase subprocess.  */
+    private final DistributedProcess<SnapshotCheckProcessRequest, HashMap<PartitionKeyV2, PartitionHashRecordV2>> phase2PartsHashes;
+
+    /** */
+    public SnapshotCheckDistributedProcess(GridKernalContext kctx) {
+        this.kctx = kctx;
+
+        log = kctx.log(getClass());
+
+        phase1CheckMetas = new DistributedProcess<>(kctx, SNAPSHOT_CHECK_METAS, this::prepareAndCheckMetas,
+            this::reducePreparationAndMetasCheck);
+
+        phase2PartsHashes = new DistributedProcess<>(kctx, SNAPSHOT_VALIDATE_PARTS, this::validateParts,
+            this::reduceValidatePartsAndFinish);
+
+        kctx.event().addLocalEventListener((evt) -> nodeLeft(((DiscoveryEvent)evt).eventNode()), EVT_NODE_FAILED, EVT_NODE_LEFT);
+    }
+
+    /**
+     * Stops the process with the passes exception.
+     *
+     * @param th The interrupt reason.
+     * @param rqFilter If not {@code null}, used to filter which requests/process to stop. If {@code null}, stops all the validations.
+     */
+    public void interrupt(Throwable th, @Nullable Function<SnapshotCheckProcessRequest, Boolean> rqFilter) {
+        if (requests.isEmpty())
+            return;
+
+        requests.values().forEach(rq -> {
+            if (rqFilter == null || rqFilter.apply(rq)) {
+                rq.error(th);
+
+                clean(rq.requestId(), th, null, null);
+            }
+        });
+    }
+
+    /** Stops the related validation if the node is a mandatory one. */
+    private void nodeLeft(ClusterNode node) {
+        if (node.isClient() || requests.isEmpty())
+            return;
+
+        interrupt(
+            new ClusterTopologyCheckedException("Snapshot checking stopped. A node left the cluster: " + node + '.'),
+            rq -> rq.nodes().contains(node.id())
+        );
+    }
+
+    /** Phase 2 and process finish. */
+    private IgniteInternalFuture<?> reduceValidatePartsAndFinish(
+        UUID procId,
+        Map<UUID, HashMap<PartitionKeyV2, PartitionHashRecordV2>> results,
+        Map<UUID, Throwable> errors
+    ) {
+        clean(procId, null, results, errors);
+
+        return FINISHED_FUT;
+    }
+
+    /** Phase 2 beginning.  */
+    private IgniteInternalFuture<HashMap<PartitionKeyV2, PartitionHashRecordV2>> validateParts(
```

Comment:
Why does it actually `HashMap` and not just a `Map`?

Reply:
Because `DistributedProcess` requires a `Serializable` result:
`DistributedProcess<I extends Serializable, R extends Serializable>`

Reply:
Let's use DTO classes, like ValidatePartitionsResponse. It's very weird to see smth like. 

```
            if (!(locMetas instanceof ArrayList))
                locMetas = new ArrayList<>(locMetas);
...
            locMetasChkFut.onDone((ArrayList<SnapshotMetadata>)locMetas);
```

Or just use an assert `assert locMetas instanceof ArrayList` to get an exception during tests. Recreate collection isn't a right way